sls-versions/src/main/java/com/palantir/sls/versions/CompactVersion.java

```diff
@@ -0,0 +1,211 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.sls.versions;
+
+import com.google.errorprone.annotations.CompileTimeConstant;
+import com.palantir.logsafe.Preconditions;
+import com.palantir.logsafe.SafeArg;
+import com.palantir.logsafe.exceptions.SafeIllegalArgumentException;
+import java.util.OptionalInt;
+
+/**
+ * Stores a compact representation of {@link OrderableSlsVersion} that is lexicographically ordered when rendered as
+ * a sequence of bytes. CompactVersion allocates 20 bits to store each of the five numeric components of an
+ * {@link OrderableSlsVersion}. To enable representing bytes as two SafeLong values, this implementation packs at most
+ * 53 bits into a single {@code long}.
+ *
+ * <p>Bits are allocated as follows, from lowest bits to highest:
+ * <code>
+ * LSB SafeLong:
+ * 20 bits: distance from release
+ *  2 bits: priority1 (1 = RELEASE_CANDIDATE_SNAPSHOT, 0 = else)
+ * 20 bits: RC number
+ *  2 bits: priority2 (2 = RELEASE_SNAPSHOT, 1 = RELEASE, 0 = RELEASE_CANDIDATE)
+ *  8 bits: lowest 8 bits of patch
+ *
+ * MSB SafeLong:
+ * 12 bits: highest 8 bits of patch
+ * 20 bits: minor
+ * 20 bits: major
+ * </code>
+ */
+public final class CompactVersion implements Comparable<CompactVersion> {
+    private static final int MASK_20_BITS = 0xFFFFF;
+    private static final int MASK_12_BITS = 0xFFF;
+    private static final int MASK_8_BITS = 0xFF;
+    private static final int MASK_4_BITS = 0x3;
+
+    private final long msb;
+    private final long lsb;
+
+    private CompactVersion(long msb, long lsb) {
+        this.msb = msb;
+        this.lsb = lsb;
+    }
+
+    public long getMsb() {
+        return msb;
+    }
+
+    public long getLsb() {
+        return lsb;
+    }
+
+    public static CompactVersion from(OrderableSlsVersion version) {
+        long patch = encode20b(version.getPatchVersionNumber(), "patch");
+
+        int rcNumber = version.firstSequenceVersionNumber().orElse(0);
+        int distanceFromVersion = version.secondSequenceVersionNumber().orElse(0);
+        if (version.getType().equals(SlsVersionType.RELEASE_SNAPSHOT)) {
+            // in this version format (1.0.0-10-gaaaaaa), the first sequence number represents the distance
+            // from the version rather than the implicit RC number
+            rcNumber = 0;
+            distanceFromVersion = version.firstSequenceVersionNumber().orElse(0);
+        }
+
+        long lsb = encode20b(distanceFromVersion, "distanceFromVersion")
+                + (priority1(version.getType()) << 20)
+                + (encode20b(rcNumber, "rcNumber") << 22)
+                + (priority2(version.getType()) << 42)
+                + ((patch & 0xFF) << 44);
+        long msb = ((patch & 0xFFF00) >> 8)
```

Comment:
Haven't verified 100%, but have you convinced yourself that you won't run into ordering problems when a leading `1` bit of the msb or lsb long indicates a 2-complement negative number? (this would be fine for lexicographic comparison, but would screw up the current implementation of `compareTo` because it's in terms of longs.)

Reply:
Yes, the tests use the maximum values and ensure round-tripping

Reply:
this works "accidentally" because you only use the top 50 and 52 bits. if the 0th bit (the highest-order one) were ever `1`, then you get a negative long. maybe add to class-level docs, `Note that the correctness of this implementation relies on the fact that strictly less than 64 bits are used for the encoding, see {@link #compareTo}` and this to compareTo, `Note that the correctness of the comparison implementation relies on the fact that the msb and lsb never have a leading 1 because we only use strictly less than 64 bits for the encoding. Without this constraint, a leading 1 in the (two-complement) msb or lsb would would represent a negative number and thus render this compareTo implementation incorrect.`

Reply:
chose slightly different text, but added a class note.

also noting that there are tests that ensure the value does not exceed 53 bits in the maximum case, and that values that would require more than 20 bits of storage throw, the combination of which should test this implementation detail.

Reply:
yeah, agree the status quo is well covered, but I think in this kind of scenario it's important that we document *why* we have this constraint. otherwise, in 2 years someone may come along, add a few bits, and change the test to 64 bits.